lib/ansible/plugins/strategy/__init__.py

```diff
@@ -1148,11 +1148,38 @@ def _evaluate_conditional(h):
                 skip_reason += ', not clearing facts and fact cache for %s' % target_host.name
         elif meta_action == 'clear_host_errors':
             if _evaluate_conditional(target_host):
+                missing_restore_state = []
                 for host in self._inventory.get_hosts(iterator._play.hosts):
                     self._tqm._failed_hosts.pop(host.name, False)
-                    self._tqm._unreachable_hosts.pop(host.name, False)
+                    unreachable = self._tqm._unreachable_hosts.pop(host.name, False)
+
+                    # If we're in a nested block, the failed host may not be in TQM's _failed_hosts yet.
+                    # Check is the host is failed with the PlayIterator since it considers child tasks.
+                    failed = iterator.is_failed(host)
+
                     iterator._host_states[host.name].fail_state = iterator.FAILED_NONE
+
+                    if not (failed or unreachable):
+                        continue
+
+                    # Strategy plugins should mark hosts as failed with PlayIterator.mark_host_failed.
+                    # If it hasn't, keep a list of hosts that can't be restored to add to msg.
+                    if failed and host.restore_state is None:
+                        missing_restore_state.append(host.name)
+                        continue
+
+                    if failed:
+                        iterator._host_states[host.name] = host.restore_state
+                        host.restore_state = None
+
+                    # Iterate over tasks until we get to clear_host_errors
+                    s, t = iterator.get_next_task_for_host(host, peek=False)
+                    while t is not None and t.args.get('_raw_params') != 'clear_host_errors':
```

Comment:
There is a problem when `clear_host_errors` wouldn't be found even if it existed.

Consider following play:
```yaml
- hosts: h1, h2
  gather_facts: no
  tasks:
    - debug:

    - fail:
      when: inventory_hostname == 'h1'

    - block:
      - fail:
      rescue:
        - meta: clear_host_errors

    - debug:
```

`clear_host_errors` is executed in the rescue section on host `h2` so the last `debug` task should execute on both hosts but it is actually executed only on `h2`.

It is because only way `PlayIterator` gets into `rescue` state is by the host being marked as `failed` which is done in the strategy. Here we only iterate through the tasks without the context strategies provide. So when `PlayIterator` gets to the `fail` task in the `block..rescue` it does not know the task fails and so it skips the `rescue` section and does not find the `clear_hosts_errors`, iterating through until it reaches the end of the play.

Obviously if you replace `rescue` by `always` in the above play, it works correctly because `PlayIterator` would iterate through `always` section.

Ugh, hopefully it makes sense.

Reply:
Ugh, yes. An alternative I had considered was continuing to iterate over failed hosts so clear_host_errors could just bring them back into play where they are. It's a little more invasive, isn't automatically fixed for 3rd party strategy plugins, and I'm not sure that `clear_host_errors` could be compatible with `free` anymore.  And it doesn't solve keeping linear tasks in lockstep/duplicate task banners... Thanks so much for pondering this one. I'm not sure if there's a great solution here, but I'll keep thinking about it.